keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,504 @@
+### User Stories (Optional)
+
+#### Story 1
+
+As a machine-learning researcher, I run an indexed job which drivers (leaders) are launched
+as index=0,1,2, and workers are launched as non index=3+.
+I want to care about only the drivers when the result of job is evaluated.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    requiredIndexes: "0-2"
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 2
+
+As a machine-learning researcher, I run an indexed job with an autoscale semantic
+set as a minimum number is 5 and a maximum number is 20.
+I want to mark the job as successful at least 5 indexes succeeded, even if 
+all indexes don't succeed.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    minSucceededIndexes: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./sample"]
+```
+
+#### Story 3
+
+As a machine-learning researcher from the [Story 2](#story-2), I want to mark the job
+as successful if 50% of all indexes succeed because I can not know how many indexes 
+will have in advance.
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  successfulPolicy:
+    minSucceededIndexes: 50%
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessfulPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successfulPolicy.requiredIndexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `requiredIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +option
```

Comment:
```suggestion
	// +optional
```

Reply:
See this in a few places but it’s not a big deal.

Reply:
Done.